File: Step01_Basics/Lec6_Basic_Hashing/Program_1_umap.cpp
#include <bits/stdc++.h>
using namespace std;

// Count elements of array using unordered map

int main() {
    int n;
    cin >> n;
    int arr[n];
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }

    // precompute
    unordered_map<int, int> mpp;
    for (int i = 0; i < n; i++) {
        mpp[arr[i]]++;
    }

    // interate in the map
    cout << "MAP STRUCTURE:" << endl;
    for (auto element : mpp) {
        cout << element.first << "->" << element.second << endl;
    }
    cout << endl;

    int q;
    cin >> q;
    while (q--) {
        int num;
        cin >> num;
        // fetch
        cout << num << " appears " << mpp[num] << " times" << endl;
    }
}

// unordered_map takes time-complexity: O(1) [best/average case]
//                                      O(n) [worst case]

// use unordered_map in most cases,
// switch to map if umap performs worse

// umap does not support pair<>, vector<> as key
// umap only supports int, double, char, string

File: Step01_Basics/Lec6_Basic_Hashing/Program_1_umap.java
package Step01_Basics.Lec6_Basic_Hashing;

import java.util.*;

// Count elements of array using unordered map

public class Program_1_umap {
    public static void main(String[] args) {
        int[] arr = {1, 2, 3, 1, 2, 3, 4, 1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 6};
        int n = arr.length;

        // precompute
        Map<Integer, Integer> umap = new HashMap<>();
        for (int element : arr) {
            umap.put(element, umap.getOrDefault(element, 0) + 1);
        }

        // iterate in the map
        System.out.println("Map Structure :");
        for (var mapElement : umap.entrySet()) {
            System.out.println(mapElement.getKey() + " -> " + mapElement.getValue());
        }

        int[] query = {1, 4, 7};
        for (int num : query) {
            System.out.println(umap.getOrDefault(num, 0));
        }
    }
}

// every operation of map takes time-complexity: O(1) & O(n) [worst case]
